File: platform/src/main/java/ru/abondin/hreasy/platform/api/GlobalWebErrorsHandler.java
package ru.abondin.hreasy.platform.api;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.r2dbc.spi.R2dbcException;
import lombok.RequiredArgsConstructor;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import org.springframework.boot.web.reactive.error.ErrorWebExceptionHandler;
import org.springframework.core.Ordered;
import org.springframework.core.annotation.Order;
import org.springframework.dao.DataAccessException;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.r2dbc.BadSqlGrammarException;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.security.authentication.BadCredentialsException;
import org.springframework.security.core.AuthenticationException;
import org.springframework.security.web.server.ServerAuthenticationEntryPoint;
import org.springframework.security.web.server.authorization.ServerAccessDeniedHandler;
import org.springframework.stereotype.Component;
import org.springframework.web.server.ResponseStatusException;
import org.springframework.web.server.ServerWebExchange;
import reactor.core.publisher.Mono;
import ru.abondin.hreasy.platform.BusinessError;
import ru.abondin.hreasy.platform.I18Helper;
import ru.abondin.hreasy.platform.config.HrEasyCorsWebFilter;

import java.util.HashMap;

/**
 * Handle errors
 */
@Component
@Order(Ordered.HIGHEST_PRECEDENCE)
@RequiredArgsConstructor
@Slf4j
public class GlobalWebErrorsHandler implements ErrorWebExceptionHandler, ServerAccessDeniedHandler, ServerAuthenticationEntryPoint {
    private final HrEasyCorsWebFilter corsWebFilter;
    private final I18Helper i18Helper;
    private final ObjectMapper json;

    @Override
    public Mono<Void> commence(ServerWebExchange serverWebExchange, AuthenticationException e) {
        return handle(serverWebExchange, e);
    }

    @Override
    public Mono<Void> handle(ServerWebExchange serverWebExchange, AccessDeniedException e) {
        return handle(serverWebExchange, (Throwable) e);
    }

    @SneakyThrows(JsonProcessingException.class)
    @Override
    public Mono<Void> handle(ServerWebExchange exchange, Throwable ex) {
        var response = exchange.getResponse();
        final Object errorDto;
        if (ex instanceof BusinessError be) {
            response.setStatusCode(HttpStatus.UNPROCESSABLE_ENTITY);
            errorDto = new BusinessErrorDto(be.getCode(),
                    i18Helper.localize(be.getCode(), be.getLocalizationArgs()), be.getAttrs());
        } else if (ex instanceof BadCredentialsException be) {
            response.setStatusCode(HttpStatus.UNAUTHORIZED);
            errorDto = new BusinessErrorDto("errors.bad.credentials", i18Helper.localize("errors.bad.credentials"), new HashMap<>());
        } else if (ex instanceof AuthenticationException be) {
            response.setStatusCode(HttpStatus.UNAUTHORIZED);
            errorDto = new BusinessErrorDto("errors.not.authenticated", i18Helper.localize("errors.not.authenticated"), new HashMap<>());
        } else if (ex instanceof AccessDeniedException be) {
            log.warn("Access Denied error: {}", be.getLocalizedMessage());
            response.setStatusCode(HttpStatus.FORBIDDEN);
            errorDto = new BusinessErrorDto("errors.access.denied", i18Helper.localize("errors.access.denied"), new HashMap<>());
        } else if (ex instanceof ResponseStatusException be) {
            log.warn("Response status error: {} - {}", be.getStatusCode(), be.getReason());
            response.setStatusCode(be.getStatusCode());
            errorDto = new BusinessErrorDto("errors.response.not.found", i18Helper.localize("errors.response.not.found"), new HashMap<>());
        } else if (ex instanceof DataAccessException) {
            log.error("Database error", ex);
            response.setStatusCode(HttpStatus.UNPROCESSABLE_ENTITY);
            errorDto = new BusinessErrorDto("errors.data_access", i18Helper.localize("errors.data_access"), new HashMap<>());
        } else {
            log.error("Unknown error", ex);
            response.setStatusCode(HttpStatus.INTERNAL_SERVER_ERROR);
            errorDto = new AnyExceptionDto(ex);
        }
        corsWebFilter.setupCors(exchange);
        response.getHeaders().setContentType(MediaType.APPLICATION_JSON);
        return response.writeWith(Mono.just(response.bufferFactory().wrap(json.writeValueAsBytes(errorDto))));
    }

}

File: platform/src/main/kotlin/ru/abondin/hreasy/platform/api/GlobalWebErrorsHandler.kt
package ru.abondin.hreasy.platform.api

import com.fasterxml.jackson.databind.ObjectMapper
import org.apache.commons.lang3.exception.ExceptionUtils
import org.springframework.boot.web.reactive.error.ErrorWebExceptionHandler
import org.springframework.context.MessageSource
import org.springframework.core.Ordered
import org.springframework.core.annotation.Order
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.security.access.AccessDeniedException
import org.springframework.security.authentication.BadCredentialsException
import org.springframework.security.core.AuthenticationException
import org.springframework.security.web.server.ServerAuthenticationEntryPoint
import org.springframework.security.web.server.authorization.ServerAccessDeniedHandler
import org.springframework.stereotype.Component
import org.springframework.web.server.ResponseStatusException
import org.springframework.web.server.ServerWebExchange
import reactor.core.publisher.Mono
import ru.abondin.hreasy.platform.BusinessError
import ru.abondin.hreasy.platform.config.HrEasyCorsWebFilter
import ru.abondin.hreasy.platform.logger
import java.util.*

/**
 * Handle errors
 */
@Component
@Order(Ordered.HIGHEST_PRECEDENCE)
class GlobalWebErrorsHandler(
        val corsWebFilter: HrEasyCorsWebFilter,
        val messageSource: MessageSource,
        val json: ObjectMapper
) : ErrorWebExceptionHandler, ServerAccessDeniedHandler, ServerAuthenticationEntryPoint {

    override fun handle(exchange: ServerWebExchange, ex: Throwable): Mono<Void> {
        logger().error("Handle error", ex);
        val response = exchange.response;
        val errorDto: Any;
        when (ex) {
            is BusinessError -> {
                response.statusCode = HttpStatus.UNPROCESSABLE_ENTITY;
                errorDto = BusinessErrorDto(ex.code, localize(ex.code, ex.localizationArgs), ex.attrs);
            }
            is BadCredentialsException -> {
                response.statusCode = HttpStatus.UNAUTHORIZED;
                errorDto = BusinessErrorDto("errors.bad.credentials", localize("errors.bad.credentials"), mapOf());
            }
            is AuthenticationException -> {
                response.statusCode = HttpStatus.UNAUTHORIZED;
                errorDto = BusinessErrorDto("errors.not.authenticated", localize("errors.not.authenticated"), mapOf());
            }
            is AccessDeniedException -> {
                response.statusCode = HttpStatus.FORBIDDEN;
                errorDto = BusinessErrorDto("errors.access.denied", localize("errors.access.denied"), mapOf());
            }
            is ResponseStatusException ->{
                response.statusCode = HttpStatus.NOT_FOUND;
                errorDto = BusinessErrorDto("errors.response.not.found", localize("errors.response.not.found"), mapOf());
            }
            else -> {
                response.statusCode = HttpStatus.INTERNAL_SERVER_ERROR;
                errorDto = AnyExceptionDto(ex);
            }
        }
        corsWebFilter.setupCors(exchange);
        response.headers.contentType = MediaType.APPLICATION_JSON;
        return response.writeWith(Mono.just(response.bufferFactory().wrap(json.writeValueAsBytes(errorDto))));
    }

    override fun handle(exchange: ServerWebExchange, denied: AccessDeniedException): Mono<Void> {
        return handle(exchange, denied as Throwable);
    }

    override fun commence(exchange: ServerWebExchange, e: AuthenticationException): Mono<Void> {
        return handle(exchange, e as Throwable);
    }


    private fun localize(code: String, vararg args: Any): String {
        return messageSource.getMessage(code, args, Locale("ru"));
    }
}

data class AnyExceptionDto(val message: String) {
    constructor(ex: Throwable) : this(ExceptionUtils.getMessage(ex));
}

data class BusinessErrorDto(
        val code: String,
        val message: String,
        val attrs: Map<String, String>
) {
}